functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+````
+
+Also, for each clause, a vector with the names of the variables used in that clause is set as its metadata. 
+
+````clojure
+(defmacro clause-term-meta [clause-term]
+   (cond
+   (coll? clause-term)  (first (filter #(variable? % false) (map str clause-term))) 
+   (variable? (str clause-term) false) (str clause-term) 
+   :no-variable-in-clause nil))
+````
+
+We use *pred-clause* to iterate over the terms in each clause: 
+
+````clojure
+(defmacro pred-clause [clause]
+   (loop [[trm# & rst-trm#] clause exprs# [] metas# []]
+     (if  trm#
+          (recur rst-trm# (conj exprs# `(clause-term-expr ~ trm#)) 
+                       (conj metas#`(clause-term-meta ~ trm#)))
+          (with-meta exprs# {:db/variable metas#}))))
+````
+
+Iterating over the clauses themselves happens in *q-clauses-to-pred-clauses*:
+          
+````clojure
+(defmacro  q-clauses-to-pred-clauses [clauses]
+     (loop [[frst# & rst#] clauses preds-vecs# []]
+       (if-not frst#  preds-vecs#
+         (recur rst# `(conj ~preds-vecs# (pred-clause ~frst#))))))
+````
+We are once again relying on the fact that macros do not eagerly evaluate their arguments. This allows us to define a simpler API where users provide variable names as symbols (e.g., ````?name````) instead of asking the user to understand the internals of the engine by providing variable names as strings ( e.g., ````"?name"````), or even worse - quoting the variable name (e.g., ````'?name````) and by that ask users to understand mechanisms that are taught in Clojure's class of defense against  the dark arts.
+
+At the end of this phase, our example yields the following set for the *:find* part: 
+````clojure #{"?nm" "?bd"} ```` and this structure for the *:where* part (each cell in the _Predicate Clause_ column holds the metadata found in its neighbor at the _Meta Clause_ column):
+
+<table>
+<tr>
+	<td>Query Clause</td>
+	<td>Predicate Clause</td>
+	<td>Meta Clause</td>
+</tr>
+<tr>
+	<td>[?e  :likes "pizza"]</td>
+	<td>[#(= % %)  #(= % :likes)  #(= % "pizza")]</td>
+	<td>["?e" nil nil]</td>
+</tr>
+<tr>
+	<td>[?e  :name  ?nm]</td>
+	<td>[#(= % %)  #(= % :name) #(= % %)]</td>
+	<td>["?e" nil "?nm"]</td>
+</tr>
+<tr>
+	<td>[?e  :speak "English"]</td>
+	<td>[#(= % %) #(= % :speak) #(= % "English")]</td>
+	<td>["?e" nil nil]</td>
+</tr>
+<tr>
+	<td>[?e  :birthday (birthday-this-month? ?bd)]</td>
+	<td>[#(= % %) #(= % :birthday) #(birthday-this-month? %)]</td>
+	<td>["?e" nil "?bd"]
+</td>
+</tr>
+</table>
+
+This structure acts as the query that is executed in a later phase, once the engine decides on the right plan of execution.
+
+#### Phase 2 - Making a plan
+In this phase, we inspect the query to determine a good plan that will produce the result it describes.
+
+In general, this will involve choosing the appropriate index and constructing a plan in the form of a function.
+
+We choose the index based on the _single_ joining variable (that can operate on only one kind of elements) 

```

Comment:
Changed it to single kind of elements (as there may be many elements of the same kind)